### PR TITLE
Add CBMC proof for derive_data_key

### PR DESCRIPTION
**Pull request by adpaco-aws**
This PR adds a proof for derive_data_key, a function from session_decrypt.c, and also patches #653 for the proof to work.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

**Comment by adpaco-aws** [comment]
@feliperodri @alex-chew I have added some content to this proof so that it conforms to the style followed by previous proofs (i.e., use "ensure" functions and similar). However, in this proof we do not need a valid `aws_cryptosdk_dec_materials`, only the `unencrypted_data_key` buffer contained within has to be valid. A similar thing happens with `aws_cryptosdk_session` (we only need to have assumptions for some header buffers or alg_properties.

Finally, not sure which postconditions would be godd to check for. In case of failure, `aws_raise_errors` shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.

**Comment by alex-chew** [comment]
Mostly LGTM. I noticed some commented-out lines referencing `ensure_nondet_sig_ctx_has_allocated_members`. Does https://github.com/aws/aws-encryption-sdk-c/issues/667#issuecomment-744898996 make those lines work, or do we not need them (in which case we can simply remove them)?

> Finally, not sure which postconditions would be godd to check for. In case of failure, aws_raise_errors shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.

Can we not retrieve the raised error code using `aws_last_error`?

**Comment by adpaco-aws** [comment]
> Mostly LGTM. I noticed some commented-out lines referencing `ensure_nondet_sig_ctx_has_allocated_members`. Does [#667 (comment)](https://github.com/aws/aws-encryption-sdk-c/issues/667#issuecomment-744898996) make those lines work, or do we not need them (in which case we can simply remove them)?

We did not need them but it is good to have all members allocated. Anyway, the issue is now fixed.

> > Finally, not sure which postconditions would be godd to check for. In case of failure, aws_raise_errors shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.
> 
> Can we not retrieve the raised error code using `aws_last_error`?

Thanks for the tip! I am still not familiar with all these functions 😞 

I have added a couple post-conditions using it. More complex ones (like the ones on `aws_cryptosdk_private_derive_key`) can come from different functions and force us to reproduce part of the code in `derive_data_key` or other functions, so I do not think it is a good idea adding them.

**Comment by alex-chew** [comment]
LGTM!

> I have added a couple post-conditions using it. More complex ones (like the ones on aws_cryptosdk_private_derive_key) can come from different functions and force us to reproduce part of the code in derive_data_key or other functions, so I do not think it is a good idea adding them.

That's fair, and the ones you added look good to me.

**Review by alex-chew — APPROVED** [review]

File: verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(session->alg_props));
+    __CPROVER_assume(session->alg_props->commitment_len <= sizeof(session->key_commitment_arr));
+
+    __CPROVER_assume(aws_byte_buf_is_bounded(&session->header.message_id, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.message_id);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.message_id));
+
+    __CPROVER_assume(aws_byte_buf_is_bounded(&session->header.alg_suite_data, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.alg_suite_data);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.alg_suite_data));
```
Comment (feliperodri) [inline verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c:45]: Replace that by
```c
__CPROVER_assume(aws_cryptosdk_session_is_valid(session));
```

File: verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    __CPROVER_assume(materials != NULL);
+    __CPROVER_assume(aws_byte_buf_is_bounded(&materials->unencrypted_data_key, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&materials->unencrypted_data_key);
+    __CPROVER_assume(aws_byte_buf_is_valid(&materials->unencrypted_data_key));
```
Comment (feliperodri) [inline verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c:50]: Replace that by
```c
__CPROVER_assume(aws_cryptosdk_dec_materials_is_valid(materials));
```

File: verification/cbmc/sources/make_common_data_structures.c [hunk 3]
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
...
 }
+
+struct aws_cryptosdk_session *ensure_session_attempt_allocation(size_t max_len) {
+    struct aws_cryptosdk_session *session = malloc(sizeof(struct aws_cryptosdk_session));
+    if (session) {
+        session->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+        session->cmm         = malloc(sizeof(struct aws_cryptosdk_cmm));
+        session->header_copy = malloc(sizeof(uint8_t));
+        session->alg_props   = ensure_alg_properties_attempt_allocation(max_len);
+        session->signctx     = ensure_nondet_sig_ctx_has_allocated_members();
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:265]:
```c
    struct aws_allocator *alloc;
    struct aws_cryptosdk_cmm *cmm;
    uint8_t *header_copy;
    const struct aws_cryptosdk_alg_properties *alg_props;
    struct aws_cryptosdk_sig_ctx *signctx;
```
Makes sure all are properly allocated and, if they also have pointers, their pointers are properly allocated.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 4]
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
         vtable->name = ensure_c_str_is_allocated(max_len);
     }
 }
+
+struct aws_cryptosdk_session *ensure_session_attempt_allocation(size_t max_len) {
+    struct aws_cryptosdk_session *session = malloc(sizeof(struct aws_cryptosdk_session));
+    if (session) {
+        session->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+        session->cmm         = malloc(sizeof(struct aws_cryptosdk_cmm));
+        session->header_copy = malloc(sizeof(uint8_t));
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:263]:
```c
        session->cmm         = malloc(sizeof(*(session->cmm)));
        session->header_copy = malloc(sizeof(*(session->header_copy)));
```

File: verification/cbmc/sources/make_common_data_structures.c [hunk 5]
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
...
+}
+
+struct aws_cryptosdk_dec_materials *ensure_dec_materials_attempt_allocation() {
+    struct aws_cryptosdk_dec_materials *materials = malloc(sizeof(struct aws_cryptosdk_dec_materials));
+    if (materials) {
+        materials->alloc   = nondet_bool() ? NULL : can_fail_allocator();
+        materials->signctx = ensure_nondet_sig_ctx_has_allocated_members();
+    }
+    return materials;
+}
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:277]: Same here.

File: verification/cbmc/sources/openssl/ec_override.c [hunk 6]
```diff
@@ -488,5 +488,11 @@ void write_unconstrained_data(unsigned char *out, size_t len) {
...
     // Currently we ignore the len parameter and just fill the entire buffer with unconstrained data.
     // This is fine because it is strictly more general behavior than writing only len bytes.
-    __CPROVER_havoc_object(out);
+    // __CPROVER_havoc_object(out);
+    if (len > 0) {
+        size_t index;
+        unsigned char nondet_char;
+        __CPROVER_assume(index < len);
+        out[index] = nondet_char;
+    }
```
Comment (feliperodri) [inline verification/cbmc/sources/openssl/ec_override.c:497]: Do we have a GitHub issue to track this issue? Both here and in our openssl verification model?

**Review by feliperodri — CHANGES_REQUESTED** [review]

**Review by tegansb — APPROVED** [review]
Overall, LGTM modulo the comments Felipe made about allocation. I'm going to go ahead and approve it assuming those get fixed because I don't want to prevent it getting merged because of time zones.

**Comment by tegansb** [comment]
> A similar thing happens with `aws_cryptosdk_session` (we only need to have assumptions for some header buffers or alg_properties). Moreover, session is the largest struct in these proofs and I do not think we are going to a use its validating function on other proofs.
> 
Very similar to the sign_header and priv_try_gen_key proofs. Only some of the fields need to be allocated and valid for the proofs to work. I also currently check the validity of the fields independently but will use any helper functions you introduce here if they make sense for those proofs.

File: verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c (hunk 1 above)
Comment (adpaco-aws) [reply verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c]: We do have a `session_setup` function which boths performs allocation and common assumptions.

File: verification/cbmc/sources/openssl/ec_override.c (hunk 6 above)
Comment (adpaco-aws) [reply verification/cbmc/sources/openssl/ec_override.c:497]: Yes, we do: #653 https://github.com/awslabs/aws-verification-model-for-libcrypto/issues/18

File: verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (adpaco-aws) [reply verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c]: Thanks, I have included the functions used in `hdr_setup` in the new `session_setup` method.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 7]
```diff
@@ -79,13 +80,9 @@ void ensure_record_has_allocated_members(struct aws_cryptosdk_keyring_trace_reco
...
     /* iterate over each record in the keyring trace */
-    size_t num_records = aws_array_list_length(trace);
-    for (size_t idx = 0; idx < num_records; ++idx) {
-        struct aws_cryptosdk_keyring_trace_record *record;
-        if (!aws_array_list_get_at_ptr(trace, (void **)&record, idx)) {
-            /* make sure each record is valid */
-            ensure_record_has_allocated_members(record, max_len);
-        }
+    for (size_t i = 0; i < trace->length; ++i) {
+        struct aws_cryptosdk_keyring_trace_record *data = (struct aws_cryptosdk_keyring_trace_record *)trace->data;
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:84]: Could you move the declaration outside of the loop?

**Review by feliperodri — APPROVED** [review]
LGTM modulo one comment.

File: verification/cbmc/sources/make_common_data_structures.c (hunk 7 above)
Comment (adpaco-aws) [reply verification/cbmc/sources/make_common_data_structures.c]: Done!

**Pull request merged**